Add workaround for files with permissive binary licenses

harmut01 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
Fixes #13840 

Add function for checking that files with Permissive Binary Licenses are not erroneously flagged as missing permissive licenses. Scancode does not currently have our license in its database as PBL is not a recognized SPDX license. This is a temporary solution until we are able to get the license into a scancode toolkit release. 
- adds function has_binary_license to determine if a file has a PBL license 
- changes condition in SPDX check branch to ensure that files with PBL are analysed, eventually marked as SPDX and not non-permissive-licenses  
#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
Affects CI, files with PBL licenses 
#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
None
### Documentation <!-- Required -->
None 
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------
@0xc0170 @hugueskamba @ARMmbed/mbed-os-maintainers

harrisonmutai-arm committed:
Add workaround for files with permissive binary licenses

Add function "has_binary_license" to check if a file has a non-permissive license contains one. PBL is not recognized by scancode,
causing it to be flagged as a non-permissive license. CI doesn't allow
any non-permissive licenses, although, files flageed as SPDX are
allowed. Workaround causes all files with a valid PBL to be flagged as
missing an SPDX.

Add condition in "has_spdx_text_in_scancode_output" to ignore any spdx
identifier with "unknown" in the name. Scancode erroneously matches PBL to matched_rule.identifer "spdx-license-identifier: unknown-spdx". This prevents the workaround from working.

0xc0170 commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -129,8 +137,9 @@ def license_check(scancode_output_path):
-
-            if not has_spdx_text_in_analysed_file(scanned_file_content):
+            # temporary workaround for files with Permissive Binary Licenses
+            # TODO raise issue in scancode repo to get PBL added to licenses
```
can we remove TODO (either add the url or no mention of it here)

hugueskamba commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -65,7 +65,10 @@ def has_permissive_text_in_scancode_output(scancode_output_data_file_licenses):
+        ('spdx' in scancode_output_data_file_license['matched_rule']['identifier']) and not ('unknown' in
+                                                                                             scancode_output_data_file_license[
+                                                                                                 'matched_rule'][
+                                                                                                 'identifier'])
```
```suggestion
        ('spdx' in scancode_output_data_file_license['matched_rule']['identifier'])
        and not ('unknown' in scancode_output_data_file_license['matched_rule']['identifier'])
```

hugueskamba commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -129,8 +137,8 @@ def license_check(scancode_output_path):
-
-            if not has_spdx_text_in_analysed_file(scanned_file_content):
+            # temporary workaround for files with Permissive Binary Licenses
+            if not has_spdx_text_in_analysed_file(scanned_file_content) or has_binary_license(scanned_file_content):
```
```suggestion
            if (
                    not has_spdx_text_in_analysed_file(scanned_file_content)
                    or has_binary_license(scanned_file_content)
            ):
```
Am I reading this right? You are marking the file as an offender if it has a binary license?

harmut01 replied:
This has changed in my latest commit, agreed with @0xc0170 to have it check for PBLs in the function has_permissive_text_in_scancode_output.

hugueskamba commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -65,7 +65,10 @@ def has_permissive_text_in_scancode_output(scancode_output_data_file_licenses):
 def has_spdx_text_in_scancode_output(scancode_output_data_file_licenses):
     """Returns true if at least one license in the scancode output has the spdx identifier."""
```
Can you please amend the docstring to include what you just added?

hugueskamba commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -81,10 +95,11 @@ def test_missing_license_permissive_license_and_spdx(self, create_scanned_files)
             test5.h: Missing `spdx` in match.identifier but found in file tested by ScanCode. (error count += 0)
+            test6.h: Matching 'unknown-spdx' in match.identifier and Permissive Binary License in header (error count += 1)
             @inputs scancode_test/scancode_test_2.json
             @output 3
```
Please update this.
```suggestion
            @output 4
```

hugueskamba commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -75,6 +76,18 @@ def has_spdx_text_in_analysed_file(scanned_file_content):
     return bool(re.findall("SPDX-License-Identifier:?", scanned_file_content))
 
 
+def has_binary_license(scancode_output_data_file):
```
One comment I received in one of my PR was that the method was doing too much. I think it also applies to this one. Perhaps it should not be opening the file. It should just check if an opened file has a binary license.

harmut01 replied:
Not sure I agree that its doing too much, perhaps this is more a question of style. @0xc0170 what are your views on this?

hugueskamba replied:
> I'm concerned this function is doing too much. We should just open the file in the parent context and pass the file text in here. I'd be surprised if a function that's just supposed to be checking something and returning a boolean started raising exceptions on me.

https://github.com/ARMmbed/mbed-os/pull/13745#discussion_r501941046

0xc0170 replied:
@harmut01 lets put this into own function (open a file), it is used in another place as well. This would simplify the function

hugueskamba replied:
It think it is just ok to open the file in the parent context and pass it to whichever methods needs it.

hugueskamba commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -54,12 +54,13 @@ def path_leaf(path):
 
-def has_permissive_text_in_scancode_output(scancode_output_data_file_licenses):
-    """Returns true if at list one license in the scancode output is permissive."""
+def has_permissive_text_in_scancode_output(scancode_output_data_file):
```
I think you should have a function do just one thing.
the check for binary license should probably be done where `has_permissive_text_in_scancode_output` is also called.

hugueskamba commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -75,6 +76,18 @@ def has_spdx_text_in_analysed_file(scanned_file_content):
+            scanned_file_content = read_file.read()
+        return bool(re.findall("Permissive Binary License", scanned_file_content))
+    except UnicodeDecodeError:
+        userlog.warning("Unable to look for PBL text in `{}`:".format(file_path))
```
I think the recommendation for logging functions is 
`Use % formatting in logging functions and pass the % parameters as arguments`

hugueskamba commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -112,7 +125,7 @@ def license_check(scancode_output_path):
             continue
 
-        if not has_permissive_text_in_scancode_output(scancode_output_data_file['licenses']):
+        if not has_permissive_text_in_scancode_output(scancode_output_data_file):
```
This was probably ok as it was.
You probably want to do something like:
```suggestion
        if (
            not has_permissive_text_in_scancode_output(scancode_output_data_file['licenses'])
            or not has_binary_license(scanned_file_content)
        ):
```

harrisonmutai-arm committed:
Refactor workaround for scancode evaluation of PBL

Add function "has_binary_license" to check if a file has a
non-permissive license contains one. PBL is not recognized by scancode,
causing it to be flagged as a non-permissive license. CI doesn't allow
any non-permissive licenses, although, files flageed as SPDX are
allowed. Workaround causes all files with a valid PBL to be flagged as
missing an SPDX.

Add condition in "has_spdx_text_in_scancode_output" to ignore any spdx
identifier with "unknown" in the name. Scancode erroneously matches PBL
to matched_rule.identifer "spdx-license-identifier: unknown-spdx". This
prevents the workaround from working.

hugueskamba commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -55,7 +56,8 @@ def path_leaf(path):
 def has_permissive_text_in_scancode_output(scancode_output_data_file_licenses):
-    """Returns true if at list one license in the scancode output is permissive."""
+    """Returns true if at least one license in the scancode output is permissive"""
+    # temporary workaround for files with Permissive Binary Licenses
```
Undo these changes.

hugueskamba commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -112,7 +126,8 @@ def license_check(scancode_output_path):
 
-        if not has_permissive_text_in_scancode_output(scancode_output_data_file['licenses']):
+        if not (has_permissive_text_in_scancode_output(scancode_output_data_file['licenses'])
+                or has_binary_license(scancode_output_data_file)):
```
I think we use `black` for checking code formatting. It will suggest the much more readable formatting below:
```suggestion
        if not (
            has_permissive_text_in_scancode_output(
                scancode_output_data_file["licenses"]
            )
            or has_binary_license(scancode_output_data_file)
        ):
```

hugueskamba commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -112,25 +129,22 @@ def license_check(scancode_output_path):
-        if not has_permissive_text_in_scancode_output(scancode_output_data_file['licenses']):
-            scancode_output_data_file['fail_reason'] = MISSING_PERMISSIVE_LICENSE_TEXT
-            license_offenders.append(scancode_output_data_file)
+        if not (has_permissive_text_in_scancode_output(scancode_output_data_file['licenses'])):
```
```suggestion
        if not has_permissive_text_in_scancode_output(scancode_output_data_file['licenses']):
```

hugueskamba commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -112,25 +129,22 @@ def license_check(scancode_output_path):
+        if not (has_permissive_text_in_scancode_output(scancode_output_data_file['licenses'])):
+            scanned_file_content = get_file_text(scancode_output_data_file)
+
+            if not (scanned_file_content is not None and has_binary_license(scanned_file_content)):
```
Wouldn't the following do the same?
```suggestion
            if not (scanned_file_content and has_binary_license(scanned_file_content)):
```

hugueskamba commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -112,25 +129,22 @@ def license_check(scancode_output_path):
+            scanned_file_content = get_file_text(scancode_output_data_file)
 
-            if not has_spdx_text_in_analysed_file(scanned_file_content):
+            if scanned_file_content is None:
```
```suggestion
            if not scanned_file_content:
```

hugueskamba commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -81,6 +95,7 @@ def test_missing_license_permissive_license_and_spdx(self, create_scanned_files)
             test5.h: Missing `spdx` in match.identifier but found in file tested by ScanCode. (error count += 0)
+            test6.h: Matching 'unknown-spdx' in match.identifier and Permissive Binary License in header (error count += 1)
             @inputs scancode_test/scancode_test_2.json
             @output 3
```
```suggestion
            @output 4
```
You added an extra error so how does the test still pass? Shouldn't the expected value also be updated from 3 to 4?

harmut01 replied:
The docstring isn't right, hadn't updated it after our changes. I'll change it now.

hugueskamba commented on tools/test/travis-ci/scancode_evaluate_test.py:
```diff
@@ -34,22 +34,36 @@
  * limitations under the License.\
  */"
 
+BINARY_HEADER = "/*\
```
Just to be in line with the other header constants.
```suggestion
HEADER_WITH_BINARY_LICENSE = "/*\
```
